File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def paran(self, param_comb, comb, n_open, n_close, n, l):
        if l==2*n:
            param_comb.append(comb)
            return
        
        if n_open < n:
            self.paran(param_comb, comb+'(', n_open + 1, n_close, n, l+1)
        if n_close < n_open:
            self.paran(param_comb, comb+')', n_open, n_close + 1, n, l+1)

    def generateParenthesis(self, n: int) -> List[str]:
        param_comb = []
        comb = "("
        self.paran(param_comb, comb, 1, 0, n, 1)
        return param_comb
        

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:

    void generate(int openN, int closeN, int n, string& paran, vector<string>& sol){
        // Base case
        if(openN == n && closeN == openN){
            sol.push_back(paran);
            return;
        }

        if(openN<n){
            paran.push_back('(');
            generate(openN+1, closeN, n, paran, sol);
            paran.pop_back();
        }

        if(closeN<openN){
            paran.push_back(')');
            generate(openN, closeN+1, n, paran, sol);
            paran.pop_back();
        }
        
    }

    vector<string> generateParenthesis(int n) {
        vector<string> sol;
        string paran = "";

        generate(0, 0, n, paran, sol);

        return sol;  
    }
};